{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Objectives:\n",
    "1. produce a gene occurence figure to show the distribution of genes in the pan-genome among different genomes\n",
    "2. produce a gene presence/absence table\n",
    "3. produce a list of fasta files, of which each contains proteins sequences in the same ortholog group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_OG_line(line):\n",
    "    # line is a line in output file of orthomcl. Each line is a ortholog group\n",
    "    # return og_id and genome list\n",
    "    cont = line.strip().split(':')\n",
    "    return cont[0].strip(),[item.split('_')[0] for item in cont[1].split()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_genomes_in_each_cluster(clsterfile):\n",
    "    # input clsterfile: is a cluter file generated by orthomcl\n",
    "    # return a dictionary {OG_id:list(unique genomes)}\n",
    "    # \n",
    "    og_genoms = dict()\n",
    "    \n",
    "    for line in open(clsterfile):\n",
    "        if ':' not in line: continue\n",
    "        og_id,genoms = process_OG_line(line)\n",
    "        og_genoms[og_id] = list(set(genoms))\n",
    "    print('Number of gene groups:',len(og_genoms))\n",
    "    return og_genoms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_gene_occurence(og_genoms):\n",
    "    # return a list which contains the gene occruency of each gene (group)\n",
    "    # \n",
    "    \n",
    "    # 1.get total number of genomes\n",
    "    all_genomes = []\n",
    "    for genoms in og_genoms.values(): all_genomes+=genoms\n",
    "    \n",
    "    tot_genome_num = float(len(set(all_genomes)))\n",
    "    print('Number of genomes:',tot_genome_num)\n",
    "    \n",
    "    occrs = [len(genoms)/tot_genome_num for genoms in og_genoms.values()]\n",
    "    return occrs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_gene_occurence(og_genoms,outname):\n",
    "    #  produce a histogram\n",
    "    # \n",
    "    occrs = calculate_gene_occurence(og_genoms)\n",
    "    plt.figure(figsize=(4,3))\n",
    "    plt.hist(occrs,50)\n",
    "    plt.xlabel('Frequency of occurence')\n",
    "    plt.ylabel('Count')\n",
    "    plt.yscale('log')\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(outname)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1. gene occurance "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of gene groups: 233478\n",
      "Number of genomes: 343.0\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAARgAAADQCAYAAADcQn7hAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAEftJREFUeJzt3X2wXVV9xvHvQ2JAJQQxSG1CvGECStTW6gWLtTX4wgQwxlJE4gtQqSlYGKmVEQcdpR1HbK1jobGQYowykvAy6CQlNlAkBpWXJEDkRXFiGsql1AS1cVDbCP76x14pJ4d7z9n3ZZ1z1rnPZ+ZM9l5nn31+99zc56699t7rKiIwM8thv24XYGb9ywFjZtk4YMwsGweMmWXjgDGzbBwwZpaNA8bMsnHAmFk2Dhgzy2ZqtwsYj5kzZ8bAwEC3yzCbdLZs2fJERBzabruiA2ZgYIDNmzd3uwyzSUfSI3W2K/IQSdIiSct3797d7VLMrIUiAyYi1kbE0hkzZnS7FDNrociAMbMyOGDMLBsHjJllU/RZpLoGLrppn/Udl57cpUrMJhf3YMwsmyIDxqepzcpQZMD4NLVZGYoMGDMrgwPGzLJxwJhZNg4YM8vGAWNm2ThgzCybngkYSQsk3S7pCkkLul2PmY1f1oCRtELSTkkPNLUvlPSwpG2SLkrNATwJHAAM5azLzDojdw9mJbCwsUHSFGAZcCIwH1giaT5we0ScCHwEuCRzXWbWAVkDJiI2Aj9taj4W2BYR2yNiD7AaWBwRv0nP/wzYf6R9SloqabOkzbt27cpSt5lNjG6MwcwCHm1YHwJmSTpF0pXA1cA/jvTiiFgeEYMRMXjooW3nHDazLuqZ6Roi4kbgxjrbSloELJo3b17eosxsXLrRg3kMOLxhfXZqq803O5qVoRsBswk4UtJcSdOA04E1o9mBp2swK0Pu09SrgDuAl0oaknR2RDwFnAesB74PXBcRD45mv+7BmJUh6xhMRCwZoX0dsG6s+/UYjFkZeuZK3tFwD8asDEUGjJmVociA8SCvWRmKDBgfIpmVociAMbMyFBkwPkQyK0ORAeNDJLMyFBkwZlaGIgPGh0hmZSgyYHyIZFaGIgPGzMrggDGzbBwwZpZNkQHjQV6zMhQZMB7kNStDkQFjZmVwwJhZNg4YM8vGAWNm2ThgzCybIgPGp6nNylBkwPg0tVkZigwYMyuDA8bMsnHAmFk2Dhgzy8YBY2bZOGDMLJueChhJz5e0WdJbu12LmY1f1oCRtELSTkkPNLUvlPSwpG2SLmp46iPAdTlrMrPOyd2DWQksbGyQNAVYBpwIzAeWSJov6S3AQ8DOzDWZWYdMzbnziNgoaaCp+VhgW0RsB5C0GlgMHAg8nyp0fiVpXUT8Jmd9ZpZX1oAZwSzg0Yb1IeC1EXEegKSzgCdGChdJS4GlAHPmzMlbqZmNSzcCpqWIWNnm+eWSHgcWTZs27TWdqcrMxqIbZ5EeAw5vWJ+d2mrzzY5mZehGwGwCjpQ0V9I04HRgzWh24OkazMqQ+zT1KuAO4KWShiSdHRFPAecB64HvA9dFxIOj2a97MGZlyH0WackI7euAdWPdr6RFwKJ58+aNdRdm1gE9dSVvXe7BmJWhyIAxszLUChhJf1CnrVM8yGtWhro9mMtrtnWED5HMytBykFfSccDrgEMlfajhqYOAKTkLM7PytevBTKO6R2gqML3h8XPg1LyljcyHSGZlaNmDiYhvAd+StDIiHulQTW1FxFpg7eDg4Pu7XYuZjazudTD7S1oODDS+JiLemKMoM+sPdQPmeuAK4Crg6XzldMbARTfts77j0pO7VIlZf6sbME9FxD9lrWQUfCWvWRnqnqZeK+kDkl4s6ZC9j6yVteDT1GZlqNuDOTP9e2FDWwBHTGw5ZtZPagVMRMzNXYiZ9Z9aASPpjOHaI+IrE1uOmfWTuodIxzQsHwC8CbgH6ErAeJDXrAx1D5HOb1yXdDCwOktFNfhCO7MyjHW6hl8AHpcxs5bqjsGspTprBNVNjkfjv8BoZm3UHYP5bMPyU8AjETGUoR4z6yO1DpHSTY8/oLqT+gXAnpxFmVl/qDuj3WnA3cA7gNOAuyR5ugYza6nuIO/FwDERcWZEnEH196U/nq+s1nyrgFkZ6gbMfhGxs2H9J6N4rZlNUnUHef9V0npgVVp/J+P4u0ZmNjm0m5N3HnBYRFwo6RTg9empO4Cv5i7OzMrWrgfzeeCjABFxI3AjgKRXpucWZa3OzIrWbhzlsIi4v7kxtQ1kqcjM+ka7gDm4xXPPnchCzKz/tAuYzZKedUOhpD8DtkxkIZKOlnSFpBsknTuR+zaz7mg3BnMB8DVJ7+aZQBmk+ntJf9xu55JWAG8FdkbEKxraFwL/QHVf01URcWlEfB84R9J+VNNAdGwOYE8CbpZHyx5MRPw4Il4HXALsSI9LIuK4iPivGvtfCSxsbJA0BVgGnAjMB5ZImp+eextwEz4FbtYX6s4Hcxtw22h3HhEbJQ00NR8LbIuI7QCSVgOLgYciYg2wRtJNwDWjfT8z6y11L7SbSLOARxvWh4DXSloAnALsT4sejKSlwFKAOXPm5KvSzMatGwEzrIjYAGyosd1yYDnA4OBgtNnczFrIPf7YjfuJHgMOb1ifndpq893UZmXoRg9mE3CkpLlUwXI68K7R7CD3nLw+q2Q2MbIGjKRVwAJgpqQh4BMR8UVJ5wHrqU5Tr4iIB0e5347+VQEHjtnYZA2YiFgyQvs6xnEq2n9VwKwMRc7p4jEYszIUGTCe0c6sDEUGjHswZmUoMmDcgzErQ5EBY2ZlKDJgfIhkVoYiA8aHSGZl6Jl7kcwsv+aLRnMrsgdjZmUosgfT6VsFmg33W8C3D5g9W5E9GI/BmJWhyB5MiXzDpHVDp8dcmjlgMun2N9asFzhgJshoA8U9GpsMigyYbg/ymvWqXus5e5DXzLIpMmDMrAxFHiL1I4/J2Fj02iFRM/dgzCwb92DMCtLrPZZm7sGYWTZFBozngzErQ5GHSJPxz5Z4ENhKVGTAmAPHyuCA6VGlDeaZDafIMRgzK4MDxsyy8SGSWYfUGTfrt0NjB4xZl/RbmAynpwJG0tuBk4GDgC9GxM1dLsnMxiH7GIykFZJ2SnqgqX2hpIclbZN0EUBEfD0i3g+cA7wzd21mllcnBnlXAgsbGyRNAZYBJwLzgSWS5jds8rH0vJkVLHvARMRG4KdNzccC2yJie0TsAVYDi1X5DPCNiLhnuP1JWipps6TNu3btylu8mY1Lt8ZgZgGPNqwPAa8FzgfeDMyQNC8irmh+YUQsB5YDDA4ORgdqNQPanwWaDIO2o9VTg7wRcRlwWbvtPCevWRm6FTCPAYc3rM9ObbVMxpsdrfe4x9JetwJmE3CkpLlUwXI68K66L3YP5tl88+P4+TOceNkDRtIqYAEwU9IQ8ImI+KKk84D1wBRgRUQ8WHef7sFYJ7iHMn7ZAyYilozQvg5YN5Z9ugdjVoaeGuStyz2Y7stxODHaszQ+hOl9RQaMtecfxmfzZ9J5RU7X4Dl5zcpQZA/Gh0ijNxl+e4/2a/Qgbn5FBoxNDu0CYLzPW35FBozPInWef1htLIoMGB8ijd9E31czGQ7BbPSKDBgrk3tBk0+RAeNDpN7nMDEo9DR1RKyNiKUzZszodilm1kKRAWNmZSjyEMkmng9pLAf3YMwsmyIDxrcKmJWhyIDxIK9ZGYoMGDMrgwPGzLJxwJhZNooo908LSdoFPFJj05nAE5nLGY9erw9c40To9fqgfo0viYhD221UdMDUJWlzRAx2u46R9Hp94BonQq/XBxNfow+RzCwbB4yZZTNZAmZ5twtoo9frA9c4EXq9PpjgGifFGIyZdcdk6cGYWRc4YMwsm74JGEkLJT0saZuki4Z5fn9J16bn75I00IM1fkjSQ5K+J+lWSS/ptRobtvsTSSGpo6dd69Qn6bT0OT4o6ZpO1lenRklzJN0m6d70vT6pw/WtkLRT0gMjPC9Jl6X6vyfp1WN+s4go/gFMAX4EHAFMA7YC85u2+QBwRVo+Hbi2B2s8HnheWj63F2tM200HNgJ3AoO9VB9wJHAv8IK0/qJe+wypBlLPTcvzgR0drvGPgFcDD4zw/EnANwABvw/cNdb36pcezLHAtojYHhF7gNXA4qZtFgNfTss3AG+SpF6qMSJui4hfptU7gdkdrK9WjcnfAJ8B/qeTxVGvvvcDyyLiZwARsbMHawzgoLQ8A/jPDtZHRGwEftpik8XAV6JyJ3CwpBeP5b36JWBmAY82rA+ltmG3iYingN3ACztSXdP7J8PV2Ohsqt8indS2xtRdPjwiujEFXp3P8CjgKEnfkXSnpIUdq65Sp8ZPAu+RNASsA87vTGm1jfb/6og8ZWYPkvQeYBB4Q7draSRpP+BzwFldLqWVqVSHSQuoeoAbJb0yIv67q1XtawmwMiL+XtJxwNWSXhERv+l2YROtX3owjwGHN6zPTm3DbiNpKlXX9Ccdqa7p/ZPhakTSm4GLgbdFxP92qLa92tU4HXgFsEHSDqrj8zUdHOit8xkOAWsi4tcR8e/AD6kCp1Pq1Hg2cB1ARNwBHEB1k2GvqPV/tZZODi5lHLSaCmwH5vLMwNrLm7b5C/Yd5L2uB2v8PaoBwiN79XNs2n4DnR3krfMZLgS+nJZnUnX1X9hjNX4DOCstH001BqMOf68HGHmQ92T2HeS9e8zv08kvKvMHdhLVb6sfARentr+m6glA9VviemAbcDdwRA/W+G/Aj4H70mNNr9XYtG1HA6bmZyiqw7iHgPuB03vtM6Q6c/SdFD73ASd0uL5VwOPAr6l6fGcD5wDnNHyGy1L994/ne+xbBcwsm34ZgzGzHuSAMbNsHDBmlo0DxsyyccCYWTYOmB4j6WlJ9zU8BrpdUydIWpXu3P3LbtdiE8enqXuMpCcj4sAWz0+N6l6qviHpt4BvR8S8LtbQd59rL3APpgCSzpK0RtI3gVtT24WSNqXf+pc0bHuxpB9K+nbqFXw4tW/Ye0m/pJnpUn8kTZH0dw37+vPUviC95gZJP5D01b13n0s6RtJ3JW2VdLek6ZI2SnpVQx3flvS7TV/HAZK+JOn+NBfK8empm4FZqcf2h02vGZD0zYY5cuak9sMkfS3VsFXS61L7GWnbrZKuTm0rJZ3asM8nG77G2yWtobowD0nvSV/TfZKulDRl72skfSrt905Jh7WpY9j9TDqdvsrRj7ZXWT7NM1fyfi21nUV1xeUhaf0EqjlFRPVL4l+o5vh4DdWVl8+jmg5gG/Dh9JoNpCsyqS6h35GWlwIfS8v7A5upLnNfQHXH+ez0HncAr6e6/H07cEx6zUFUl8efCXw+tR0FbB7ma/srYEVafhnwH1RXWA8w8mXra4Ez0/L7gK+n5WuBC9LyFKp7y15OdQXtzNS+9/NaCZzasM8n078LgF8Ac9P60en9npPWvwCckZYDWJSW/7bhMxuujhH3M9kevpu69/wqIl41TPstEbF3Do8T0uPetH4g1Q1906lC6ZcA6TdzOycAv9PwG35G2tceqntQhtK+7qMKgt3A4xGxCSAifp6evx74uKQLqYJg5TDv9Xrg8vS6H0h6hCqMft6ivuOAU9Ly1VQ/3ABvBM5I+3oa2C3pDOD6iHgitbea82Svu6O6KRLgTVQhvSl11p4L7J1PZg9VkANsAd7Soo73ttjPpOKAKccvGpYFfDoirmzcQNIFLV7/FM8cEh/QtK/zI2J9074WAI13cz9Ni/8vEfFLSbdQTVZ0GtUPWK/4/689TTkxreG55s/1yxHx0WH28etI3RHafBZt9jOpeAymTOuB90k6EEDSLEkvoprG8u2SnitpOrCo4TU7eOaH/tSmfZ0r6TlpX0dJen6L934YeLGkY9L209P0FwBXAZcBmyLNKNfkduDde98HmJP218p3qe5+J7329rR8K9W0onvHkWYA3wTeIemFqf2QtO0Onvna3wY8Z4T3uhU4NX2WSDpE7edFHq6OseynLzlgChQRNwPXAHdIup9qCtDpEXEP1ZjAVqrb7Tc1vOyzVEFyL/vOPXIV1QDnPaomgb6S1j2VPcA7gcslbQVuIfWIImIL1eHOl0Z4+ReA/VLN11JNWdBuzpvzgT+V9D3gvcAHU/sHgePTvrZQzXv7IPAp4Fupts+lbf8ZeENqO459ey2NX9tDwMeAm9P73QK0mypyuDrGsp++5NPUfUzSJ6kGND/boff7barB5JdFH87OZqPnHoxNiDTAehfV/CcOFwPcgzGzjNyDMbNsHDBmlo0DxsyyccCYWTYOGDPL5v8AaMb7YYNnH1QAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 288x216 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "og_genoms = load_genomes_in_each_cluster('../data/orthomcl_output/orthomcl_clusters.txt')\n",
    "plot_gene_occurence(og_genoms,'../figures/pan_genome_gene_occurence.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2. gene presence absence table\n",
    "Index: species name  \n",
    "Columns: S228C genes. For those ones without S228C genes, use ortho ID\n",
    "For those orthogroups with multiple genes from S288C, randomly choose one"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_idmap(idmapfile):\n",
    "    # idmapfile: orthomcl_SeqIDs_index.txt\n",
    "    gid2orgid = dict() \n",
    "    for line in open(idmapfile): \n",
    "        cont = line.strip().split(':')\n",
    "        gid2orgid[cont[0].strip()] = cont[1].strip()\n",
    "    print(list(gid2orgid.items())[:10])\n",
    "    print('Number of genes:',len(gid2orgid))\n",
    "    return gid2orgid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def produce_gene_pa_table(orthofile,idmapfile,outname):\n",
    "    # orthofile: orthomcl_clusters.txt\n",
    "    \n",
    "    gid2orgid = load_idmap(idmapfile)\n",
    "    \n",
    "    org_ref = dict() # {(org,ref_id):True}\n",
    "    orgs = dict()\n",
    "    refs = dict()\n",
    "    for group in open(orthofile):\n",
    "        cont = group.strip().split(':')\n",
    "        orthoid = cont[0].strip()\n",
    "        genes = [item for item in cont[1].split()]\n",
    "        \n",
    "        # find one reference gene\n",
    "        ref = orthoid\n",
    "        for gene in genes:\n",
    "            org = gid2orgid[gene].split('@')[0]\n",
    "            orgs[org] = True\n",
    "            if org == 'Saccharomyces_cerevisiae': \n",
    "                ref = gid2orgid[gene].split('@')[1]\n",
    "                break\n",
    "        refs[ref] = True\n",
    "        for gene in genes:\n",
    "            org = gid2orgid[gene].split('@')[0]\n",
    "            org_ref[(org,ref)] = True\n",
    "    print('Number of gene pairs:',len(org_ref))\n",
    "    # produce a tsv file\n",
    "    index = list(orgs.keys())\n",
    "    index.sort()\n",
    "    \n",
    "    columns = list(refs.keys())\n",
    "    columns.sort()\n",
    "    \n",
    "    data = np.zeros((len(index),len(columns)))\n",
    "    for i,org in enumerate(index):\n",
    "        for j,ref in enumerate(columns):\n",
    "            if org_ref.get((org,ref),False): data[i,j] = 1\n",
    "    \n",
    "    df = pd.DataFrame(data=data,index=index,columns=columns,dtype=int)\n",
    "    print(df.shape)\n",
    "    df.to_csv(outname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "def produce_one_fasta_for_each_gene_cluster(fafile,orthofile,idmapfile,outdir):\n",
    "    # load seqs\n",
    "    seqs = SeqIO.to_dict(SeqIO.parse(fafile,'fasta'))\n",
    "    gid2orgid = load_idmap(idmapfile)\n",
    "    for group in open(orthofile):\n",
    "        cont = group.strip().split(':')\n",
    "        orthoid = cont[0].strip()\n",
    "        genes = [item for item in cont[1].split()]\n",
    "        \n",
    "        if len(genes)<2: continue\n",
    "        fhand = open(os.path.join(outdir,'{0}.fasta'.format(orthoid)),'w')\n",
    "        sub_seqs = [seqs[gid2orgid[gene]] for gene in genes]\n",
    "        SeqIO.write(sub_seqs,fhand,'fasta')\n",
    "        fhand.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('186_7100', 'yHMPu5000034631_Martiniozyma_abiesophila@Seq_7101'), ('272_4809', 'yHMPu5000035048_Barnettozyma_salicaria@Seq_4810'), ('84_6277', 'Metschnikowia_matae@Seq_6278'), ('74_1964', 'Metschnikowia_continentalis@Seq_1965'), ('49_2508', 'Kazachstania_naganishii@Seq_2509'), ('339_4404', 'yHMPu5000041862_Candida_golubevii@Seq_4405'), ('79_1102', 'Metschnikowia_hibisci@Seq_1103'), ('21_2590', 'Candida_infanticola@Seq_2591'), ('278_782', 'yHMPu5000035268_Wickerhamomyces_hampshirensis@Seq_783'), ('277_5417', 'yHMPu5000035261_Candida_ponderosae@Seq_5418')]\n",
      "Number of genes: 2012541\n",
      "Number of gene pairs: 1834191\n",
      "(343, 233478)\n"
     ]
    }
   ],
   "source": [
    "produce_gene_pa_table('../data/orthomcl_output/orthomcl_clusters.txt',\n",
    "                      '../data/orthomcl_output/orthomcl_SeqIDs_index.txt',\n",
    "                      '../data/orthomcl_output/gene_pa_table.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('186_7100', 'yHMPu5000034631_Martiniozyma_abiesophila@Seq_7101'), ('272_4809', 'yHMPu5000035048_Barnettozyma_salicaria@Seq_4810'), ('84_6277', 'Metschnikowia_matae@Seq_6278'), ('74_1964', 'Metschnikowia_continentalis@Seq_1965'), ('49_2508', 'Kazachstania_naganishii@Seq_2509'), ('339_4404', 'yHMPu5000041862_Candida_golubevii@Seq_4405'), ('79_1102', 'Metschnikowia_hibisci@Seq_1103'), ('21_2590', 'Candida_infanticola@Seq_2591'), ('278_782', 'yHMPu5000035268_Wickerhamomyces_hampshirensis@Seq_783'), ('277_5417', 'yHMPu5000035261_Candida_ponderosae@Seq_5418')]\n",
      "Number of genes: 2012541\n"
     ]
    }
   ],
   "source": [
    "outdir = '../data/orthomcl_output/gene_clusters'\n",
    "if not os.path.exists(outdir): os.mkdir(outdir)\n",
    "produce_one_fasta_for_each_gene_cluster('../data/orthomcl_output/343taxa_proteins.fasta',\n",
    "                                        '../data/orthomcl_output/orthomcl_clusters.txt',\n",
    "                                        '../data/orthomcl_output/orthomcl_SeqIDs_index.txt',\n",
    "                                        outdir)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
